### PR TITLE
semant check builtin input

### PR DESCRIPTION
**Pull request by rxie25**
working on adding semantic check for builtin input.

### CODE REVIEW

**Comment by rxie25** [comment]
@jzacsh take a look at this. It's done. Let me know what you think. We were gonna work on multi-line, but too many issues with the stuff currently at hand.  We will try to do multi-line on Friday.

File: src/codegen.ml [hunk 1]
```diff
@@ -94,17 +94,29 @@ let translate sast =
...
+      if (List.mem "#include <stdio.h>\n" head_lib) then head_lib else "#include <stdio.h>\n"::head_lib
+    in
     let header head elem=
       match elem with
-      |"%" -> if (List.mem "#include <math.h>\n" head) then head else "#include <math.h>\n"::head 
-      |"^" -> if (List.mem "#include <math.h>\n" head) then head else "#include <math.h>\n"::head 
-      |"|" -> if (List.mem "#include <math.h>\n" head) then head else "#include <math.h>\n"::head 
-      |"print" -> if (List.mem "#include <stdio.h>\n" head) then head else "#include <stdio.h>\n"::head
-      |_ -> head 
+      |"%" -> add_math_lib head
```
Comment (jzacsh) [inline src/codegen.ml]: nit: maybe combine these where it makes things more readable? eg: maybe rewrite `header` to `add_lib_for_dep`:

```ocaml
let add_lib_for_dep head = function
  | _ -> head
  | "%" | "^" | "|" | "cos" | "sin" | "tan" | "sqrt" | "log" -> add_math_lib head
  | "print" -> add_stdio_lib head
in List.fold_left add_lib_for-dep [] liblist
```

File: src/semant.ml [hunk 2]
```diff
@@ -40,67 +40,125 @@ let check (contexts, finds) =
...
                |_ -> (add_lib_expre lis left)@(add_lib_expre lis right)@lis )
         | A.Unop(op, expr) -> (
             match op with
             |A.Abs -> "|"::lis
             |_ -> add_lib_expre lis expr )
+        | A.Assign(left, expr) -> add_lib_expre lis expr
         | A.Builtin(name, expr) -> (
             match name with
-            |"print" -> "print"::(List.fold_left add_lib_expre lis expr)
+            | "cos" -> "cos"::lis
```
Comment (jzacsh) [inline src/semant.ml]: _(same comment here about consolidating)_

File: src/semant.ml [hunk 3]
```diff
@@ -40,67 +40,125 @@ let check (contexts, finds) =
...
+      | "sin" -> fail ("illegal argument for sin, " ^ quot str)
+      | "cos" -> fail ("illegal argument for cos, " ^ quot str)
+      | "tan" -> fail ("illegal argument for tan, " ^ quot str)
+      | "log" -> fail ("illegal argument for log, " ^ quot str)
+      | "sqrt" -> fail ("illegal argument for sqrt, " ^ quot str)
+      | _ -> fail ("unknown built-in function, " ^ quot name)
+
+  and check_builtin_other s hd =
+      match s, hd with
+          | s, A.Assign(left, expr) -> check_builtin_name s left
```
Comment (jzacsh) [inline src/semant.ml]: maybe rename `check_builtin_name`? You're not using to `check`, but instead calling it when you *know* you're failing right?

File: src/semant.ml [hunk 4]
```diff
@@ -40,67 +40,125 @@ let check (contexts, finds) =
...
+    match expr with
+    |A.Builtin(name, expr) -> (
+      match name with
+      | "print" -> fail("Illegal use of operator on print, " ^ quot op)
+      | _ -> ()
+      )
+    | _ -> ()
+
+  and check_builtin_name name str =
+    match name with
```
Comment (jzacsh) [inline src/semant.ml]: could you DRY up this code a tiny bit? eg:

```ocaml
and check_builtin_name name str =
  let fail_ilegal_arg builtin arg = fail ("illegal argument for " ^ builtin ^ ", " ^ quot arg) in
  match name with
    | "sin" | "cos" | "tan" | "log" | "sqrt" -> fail_illegal_arg name str
    | _ -> fail ("unknown ...")
```

**Comment by rxie25** [comment]
I addressed all your comments, anything else? @jzacsh

File: src/semant.ml [hunk 5]
```diff
@@ -116,39 +112,31 @@ let check (contexts, finds) =
...
       | _ -> fail ("unknown built-in function, " ^ quot name)
 
   and check_builtin_other s hd =
       match s, hd with
-          | s, A.Assign(left, expr) -> check_builtin_name s left
-          | s, A.Strlit(str) -> check_builtin_name s str
-          | "cos", _ -> ()
-          | "sin", _ -> ()
+          | s, A.Assign(left, expr) -> match_builtin_name s left
+          | s, A.Strlit(str) -> match_builtin_name s str
```
Comment (jzacsh) [inline src/semant.ml]: at this point, you're explicitly sure that there's a semantic error, right? So `match_builtin_name` could be renamed to like... `fail_illegal_builtin_arg` or something? My point is just that it's confusing code to read. Generally I think:

- `check` means you'll make some assertion (eg: is `1 < 0` then `panic`)
- `match` means you'll pick some data structure apart and return something new

But these are just helpers that format `panic` messages for you, so maybe just rename them to not suggest the above two categories.

**Comment by jzacsh** [comment]
looks great. i'm still confused by the names used for some stuff (and left comments for it), but otherwise feel free to merge once you've resolved!

File: src/codegen.ml [hunk 6]
```diff
@@ -155,15 +155,22 @@ let translate sast =
...
   let lib=
-    let header head elem=
+    let add_math_lib head_lib =
+      if (List.mem "#include <math.h>\n" head_lib) then head_lib else "#include <math.h>\n"::head_lib
+    in
+    let add_stdio_lib head_lib =
+      if (List.mem "#include <stdio.h>\n" head_lib) then head_lib else "#include <stdio.h>\n"::head_lib
+    in
+
+    let add_lib_for_dep head elem =
```
Comment (jzacsh) [inline src/codegen.ml]: _(optional)_: can just write `let add_lib_for_dep head = function` and delete yet another line :+1:

File: src/tests/fail-cos-illegal-arg-assign.err [hunk 7]
```diff
@@ -0,0 +1 @@
+Fatal error: exception Failure("illegal argument for cos, "a"")
```
Comment (jzacsh) [inline src/tests/fail-cos-illegal-arg-assign.err]: @rxie25 The illegal argument is an assignment, right? Not the `a`? Can you fix the error message? This sounds like we disallow variables

**Pull request merged**